### PR TITLE
🚧Improve layered rendering

### PR DESCRIPTION
**Pull request by CalvinWilkinson**
<h1 style="font-weight:bold" align="center">Preview Feature Pull Request</h1>
<h2 style="font-weight:bold" align="center">📋Review Checklist📋</h2>

- [x] The **_head(source)_** branch for this pull request is a **_preview feature_** branch, with the correct naming syntax, in the following example:
  - A **_head(source)_** branch is the branch being merged into a target branch.
    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/feature/<issue-num>-<description>
      Example: preview/feature/123-my-preview-feature
      ```
    </details>

- [x] The **_head(source)_** branch for this pull request is created from a **_preview release_** branch, with the correct naming syntax, in the following example:
  <details closed><summary>Example</summary>

    ``` xml
    Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
    Example: preview/v1.2.3-preview.4
    ```
  </details>

- [x] The **_base(target)_** branch for this pull request is a **_preview release_** branch, with the correct naming syntax, in the following example:
  - A **_base(target)_** branch is the branch that the **_head(source)_** branch is merging into.

    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
      Example: preview/v1.2.3-preview.4
      ```
    </details>

💡For more information on branching, refer to the project's [branching documentation](../../Documentation/Branching.md).

- [x] Associated issue exists and is linked to this pull request.
  - One issue per pull request.

- [x] Pull request title matches the title of the associated issue.

- [x] A `preview` label has been added to the pull request.

- [x] The labels attached to this PR match the labels attached to the associated issue.

- [x] Project assigned matches the project of the associated issue.

- [x] Milestone assigned matches the milestone of the associated issue.

- [x] My code follows the coding style of this project.
  - The style is enforced by the **_.editorconfig_** files in the project and displayed as warnings.  If there is an edge case, reach out, and let's discuss it.

- [x] I have manually tested my code changes to the best of my ability.

- [x] All tests passed locally.
  - This is required because unit tests might pass locally but not in the CI environment during the status check process or vice-versa.
  - Tests might pass on the developer's machine but not necessarily on the code reviewer's machine.
  - Status checks run for this pull request when merging into the **_preview/v\*.\*.\*-preview.\*_** branch.  These status checks run every time a change is pushed to the pull request.  These checks validate version syntax, tagging, builds, unit tests, and more.

### CODE REVIEW

File: Testing/VelaptorTests/OpenGL/Batching/TextureBatchItemTests.cs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+using System.Collections.Generic;
+using System.Drawing;
+using FluentAssertions;
+using Velaptor.Graphics;
+using Velaptor.OpenGL.Batching;
+using Xunit;
+using Xunit.Abstractions;
+
+/// <summary>
+/// Test the <see cref="TextureBatchItem"/> struct.
```
Comment (kselena) [inline Testing/VelaptorTests/OpenGL/Batching/TextureBatchItemTests.cs:16]: Refactor to "Tests"
```suggestion
/// Tests the <see cref="TextureBatchItem"/> struct.
```

File: Velaptor/Batching/BatchingManager.cs [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+    /// <inheritdoc/>
+    public void AddTextureItem(TextureBatchItem item, int layer)
+    {
+        var emptyItemIndex = this.textureItems.
+            FirstItemIndex(i => i.Item.IsEmpty());
+
+        if (emptyItemIndex == -1)
+        {
+            // TODO: Replace this code and comment once GPU adjustment feature is added
```
Comment (kselena) [inline Velaptor/Batching/BatchingManager.cs:151]: Is this TODO still needed?  Check for all TODO's...there are more. 
```suggestion
            // TODO: Replace this code and comment once GPU adjustment feature is added
```

File: Velaptor/Batching/IBatchingManager.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    /// </summary>
+    /// <param name="item">The item to add.</param>
+    /// <param name="layer">The layer to add the item on.</param>
+    void AddTextureItem(TextureBatchItem item, int layer);
+
+    /// <summary>
+    /// Adds a font glyph item to the batch.
+    /// </summary>
+    /// <param name="item">The item to add.</param>
+    /// <param name="layer">The layer to add the item on.</param>
```
Comment (kselena) [inline Velaptor/Batching/IBatchingManager.cs:25]: Grammar.  Refactor to remove the "on" at the end of the sentence
```suggestion
    /// <param name="layer">The layer to add the item.</param>
```

File: Velaptor/Graphics/RenderItemComparer.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    /// <summary>
+    /// Compares two <see cref="RenderItem{T}"/>s. Returns a
+    /// value less than zero if the <paramref name="x"/>.<see cref="RenderItem{T}.Layer"/> is less than
+    /// <paramref name="y"/>.<see cref="RenderItem{T}.Layer"/>, zero if <paramref name="x"/>.<see cref="RenderItem{T}.Layer"/>
+    /// is equal to <paramref name="y"/>.<see cref="RenderItem{T}.Layer"/>, or a value greater than zero if
+    /// <paramref name="x"/>.<see cref="RenderItem{T}.Layer"/> is greater than <paramref name="y"/>.<see cref="RenderItem{T}.Layer"/>.
+    /// </summary>
+    /// <param name="x">The left side of the comparison.</param>
+    /// <param name="y">The right side of the comparison.</param>
+    /// <returns>The numerical value representing how to sort the items <paramref name="x"/> and <paramref name="y"/>.</returns>
```
Comment (kselena) [inline Velaptor/Graphics/RenderItemComparer.cs:26]: Improve code docs.  Also, list out the number 2 as two, not the actual int

File: Velaptor/Graphics/RenderMediator.cs [hunk 5]
```diff
@@ -4,41 +4,224 @@
...
 
     /// <summary>
     /// Initializes a new instance of the <see cref="RenderMediator"/> class.
     /// </summary>
     /// <param name="reactableFactory">Creates reactables for sending and receiving notifications with or without data.</param>
-    public RenderMediator(IReactableFactory reactableFactory)
+    /// <param name="textureItemComparer">Compares 2 texture batch items for the purpose of sorting.</param>
+    /// <param name="fontItemComparer">Compares 2 font batch items for the purpose of sorting.</param>
+    /// <param name="rectItemComparer">Compares 2 rect batch items for the purpose of sorting.</param>
+    /// <param name="lineItemComparer">Compares 2 line batch items for the purpose of sorting.</param>
```
Comment (kselena) [inline Velaptor/Graphics/RenderMediator.cs:44]: Change to two, not "2".
```suggestion
    /// <param name="textureItemComparer">Compares two texture batch items for the purpose of sorting.</param>
    /// <param name="fontItemComparer">Compares two font batch items for the purpose of sorting.</param>
    /// <param name="rectItemComparer">Compares two rect batch items for the purpose of sorting.</param>
    /// <param name="lineItemComparer">Compares two line batch items for the purpose of sorting.</param>
```

File: Velaptor/Graphics/RenderMediator.cs [hunk 6]
```diff
@@ -4,41 +4,224 @@
...
+        this.fontRenderBatchReactable = reactableFactory.CreateRenderFontReactable();
+        this.rectRenderBatchReactable = reactableFactory.CreateRenderRectReactable();
+        this.lineRenderBatchReactable = reactableFactory.CreateRenderLineReactable();
+
+        this.textureItemComparer = textureItemComparer;
+        this.fontItemComparer = fontItemComparer;
+        this.rectItemComparer = rectItemComparer;
+        this.lineItemComparer = lineItemComparer;
+
+        // Set all of the layers to a default value of the max value of int
```
Comment (kselena) [inline Velaptor/Graphics/RenderMediator.cs:87]: Should this be Set or Sets? 
```suggestion
        // Sets all of the layers to a default value of the max value of int
```

File: Velaptor/Graphics/RenderMediator.cs [hunk 7]
```diff
@@ -4,41 +4,224 @@
...
+            if (totalLinesOnCurrentLayer > 0)
+            {
+                var lineLayerStart = lineItems.FirstLayerIndex(currentLayer);
+
+                this.lineRenderBatchReactable.Push(
+                    lineItems.Slice(lineLayerStart, totalLinesOnCurrentLayer),
+                    PushNotifications.RenderLinesId);
+            }
+
+            // Reset the item back to the default value
```
Comment (kselena) [inline Velaptor/Graphics/RenderMediator.cs:212]: Should this be Reset or Resets?
```suggestion
            // Resets the item back to the default value
```

File: Velaptor/Graphics/RenderMediator.cs [hunk 8]
```diff
@@ -4,41 +4,224 @@
...
+                continue;
+            }
+
+            this.allLayers.Span[layerIndex] = lineLayer;
+            layerIndex++;
+        }
+
+        this.allLayers.Span.Sort();
+
+        // Render all of the items in a coordinated fashion
```
Comment (kselena) [inline Velaptor/Graphics/RenderMediator.cs:161]: Should this be Render or Renders?
```suggestion
        // Renders all of the items in a coordinated fashion
```

File: Velaptor/Graphics/Renderers/FontRenderer.cs [hunk 9]
```diff
@@ -124,6 +131,93 @@ protected override void ShutDown()
...
+    {
+        var result = new List<FontGlyphBatchItem>();
+
+        var leftGlyphIndex = 0u;
+
+        foreach (var currentCharMetric in charMetrics)
+        {
+            textPos.X += font.GetKerning(leftGlyphIndex, currentCharMetric.CharIndex);
+
+            // Create the source rect
```
Comment (kselena) [inline Velaptor/Graphics/Renderers/FontRenderer.cs:167]: Refactor all of these single line comments (if needed) to Creates, Calculates, Adjusts etc instead of Create, Calculate, Adjust. 
```suggestion
            // Creates the source rect
```

File: Velaptor/InternalExtensionMethods.cs [hunk 10]
```diff
@@ -1479,4 +1477,109 @@ public static int IndexOf<T>(this IEnumerable<T> items, Predicate<T> predicate)
 
         return -1;
     }
+
+    /// <summary>
+    /// Returns the index of the first occurence of an item that matches the given <paramref name="predicate"/> result.
+    /// </summary>
+    /// <param name="items">The items to check.</param>
+    /// <param name="predicate">Indicates if the item index should be returned.</param>
```
Comment (kselena) [inline Velaptor/InternalExtensionMethods.cs:1485]: Grammar
```suggestion
    /// <param name="predicate">Indicates whether or not the item index should be returned.</param>
```

File: Velaptor/InternalExtensionMethods.cs [hunk 11]
```diff
@@ -1479,4 +1477,109 @@ public static int IndexOf<T>(this IEnumerable<T> items, Predicate<T> predicate)
...
+    /// <typeparam name="T">The type of item in list.</typeparam>
```
Comment (kselena) [inline Velaptor/InternalExtensionMethods.cs:1486]: Grammar
```suggestion
    /// <typeparam name="T">The type of item in the list.</typeparam>
```

File: Velaptor/InternalExtensionMethods.cs [hunk 12]
```diff
@@ -1479,4 +1477,109 @@ public static int IndexOf<T>(this IEnumerable<T> items, Predicate<T> predicate)
...
+    /// <returns>The index of the item that the <paramref name="predicate"/> returned <c>true</c> on.</returns>
```
Comment (kselena) [inline Velaptor/InternalExtensionMethods.cs:1487]: Grammar
```suggestion
    /// <returns>The index of the item that the <paramref name="predicate"/> returned <c>true</c>.</returns>
```

File: Velaptor/InternalExtensionMethods.cs [hunk 13]
```diff
@@ -1479,4 +1477,109 @@ public static int IndexOf<T>(this IEnumerable<T> items, Predicate<T> predicate)
...
+        }
+
+        return -1;
+    }
+
+    /// <summary>
+    /// Returns the index of the first item that matches the given <paramref name="layer"/>.
+    /// </summary>
+    /// <param name="items">The items to check.</param>
+    /// <param name="layer">The layer to check for.</param>
```
Comment (kselena) [inline Velaptor/InternalExtensionMethods.cs:1512]: Grammar
```suggestion
    /// <param name="layer">The layer to check.</param>
```

File: Velaptor/InternalExtensionMethods.cs [hunk 14]
```diff
@@ -1479,4 +1477,109 @@ public static int IndexOf<T>(this IEnumerable<T> items, Predicate<T> predicate)
...
+    /// <typeparam name="T">The type of <see cref="RenderItem{T}"/>.</typeparam>
+    /// <returns>
+    ///     The index of the item in the list of <paramref name="items"/>.
+    ///     <br/>
+    ///     The value of -1 will be returned if the layer can not be found.
```
Comment (kselena) [inline Velaptor/InternalExtensionMethods.cs:1517]: Spelling (cannot is one word)
```suggestion
    ///     The value of -1 will be returned if the layer cannot be found.
```

File: Velaptor/InternalExtensionMethods.cs [hunk 15]
```diff
@@ -1479,4 +1477,109 @@ public static int IndexOf<T>(this IEnumerable<T> items, Predicate<T> predicate)
...
+        }
+
+        return -1;
+    }
+
+    /// <summary>
+    /// Gets the total number of items that are on the given <paramref name="layer"/>.
+    /// </summary>
+    /// <param name="items">The items to check.</param>
+    /// <param name="layer">The layer to check for.</param>
```
Comment (kselena) [inline Velaptor/InternalExtensionMethods.cs:1536]: Refactor to omit the "for" at the end. 
```suggestion
    /// <param name="layer">The layer to check.</param>
```

File: Velaptor/OpenGL/Batching/RenderItem.cs [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+namespace Velaptor.OpenGL.Batching;
+
+/// <summary>
+/// A single batch item with associated meta-data about how it should be rendered.
+/// </summary>
+/// <typeparam name="T">The <c>struct</c> that represents the batch item.</typeparam>
+internal readonly record struct RenderItem<T>
+{
+    /// <summary>
+    /// Gets the layer that the <see cref="Item"/> should be rendered on.
```
Comment (kselena) [inline Velaptor/OpenGL/Batching/RenderItem.cs:14]: Refactor to below
```suggestion
    /// Gets the layer where the <see cref="Item"/> should be rendered.
```

**Review by kselena — CHANGES_REQUESTED** [review]

**Review by KinsonDigitalAdmin — APPROVED** [review]

**Review by kselena — APPROVED** [review]
